clippy_lints/src/derivable_impls.rs

```diff
@@ -0,0 +1,121 @@
+    /// specialized than what derive will produce. This lint can't detect the manual `impl`
+    /// has exactly equal bounds, and therefore this lint is disabled for types with
+    /// generic parameters.
+    ///
+    pub DERIVABLE_IMPLS,
+    complexity,
+    "manual implementation of the `Default` trait which is equal to a derive"
+}
+
+declare_lint_pass!(DerivableImpls => [DERIVABLE_IMPLS]);
+
+fn is_path_self(cx: &LateContext<'_>, e: &Expr<'_>, self_path: &Path<'_>) -> bool {
+    if let ExprKind::Path(QPath::Resolved(_, p)) = e.kind {
+        matches!(p.res, Res::SelfCtor(..)) || SpanlessEq::new(cx).eq_path(p, self_path)
+    } else {
+        false
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for DerivableImpls {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        if_chain! {
+            if let ItemKind::Impl(Impl {
+                of_trait: Some(ref trait_ref),
+                items: [child],
+                self_ty,
+                ..
+            }) = item.kind;
+            if let attrs = cx.tcx.hir().attrs(item.hir_id());
+            if !is_automatically_derived(attrs);
+            if !in_macro(item.span);
+            if let Some(def_id) = trait_ref.trait_def_id();
+            if cx.tcx.is_diagnostic_item(sym::Default, def_id);
+            if let impl_item_hir = child.id.hir_id();
+            if let Some(Node::ImplItem(impl_item)) = cx.tcx.hir().find(impl_item_hir);
+            if let ImplItemKind::Fn(_, b) = &impl_item.kind;
+            if let Body { value: func_expr, .. } = cx.tcx.hir().body(*b);
+            if let Ty {
+                kind: TyKind::Path(QPath::Resolved(_, self_ty_path)),
+                ..
+            } = self_ty;
+            then {
+                if let Some(generic) = self_ty_path.segments.last().and_then(|s| s.args) {
+                    for arg in generic.args {
+                        if !matches!(arg, GenericArg::Lifetime(_)) {
+                            return;
+                        }
+                    }
+                }
+                let should_emit = match remove_blocks(func_expr).kind {
+                    ExprKind::Tup(fields) => fields.iter().all(|e| is_default_equivalent(cx, e)),
+                    ExprKind::Call(callee, args)
+                        if is_path_self(cx, callee, self_ty_path) => args.iter().all(|e| is_default_equivalent(cx, e)),
+                    ExprKind::Struct(_, fields, _) => fields.iter().all(|ef| is_default_equivalent(cx, ef.expr)),
+                    _ => false,
+                };
+                if should_emit {
+                    let path_string = self_ty_path
+                        .segments
+                        .iter()
+                        .map(|segment| {
+                            segment.ident.to_string()
+                        })
+                        .collect::<Vec<_>>()
+                        .join("::");
```

Comment:
Can you use `cx.tcx.def_path_str` with the `DefId` from `cx.tcx.type_of(item.def_id).ty_adt_def()` here?